mmv1/third_party/terraform/framework_utils/framework_config.go.erb

```diff
@@ -0,0 +1,668 @@
+		}, "")
+		if customEndpoint != nil {
+			data.DataplexCustomEndpoint = types.StringValue(customEndpoint.(string))
+		}
+	}
+
+	if data.EventarcCustomEndpoint.IsNull() {
+		customEndpoint := MultiEnvDefault([]string{
+			"GOOGLE_EVENTARC_CUSTOM_ENDPOINT",
+		}, "")
+		if customEndpoint != nil {
+			data.EventarcCustomEndpoint = types.StringValue(customEndpoint.(string))
+		}
+	}
+
+	if data.FirebaserulesCustomEndpoint.IsNull() {
+		customEndpoint := MultiEnvDefault([]string{
+			"GOOGLE_FIREBASERULES_CUSTOM_ENDPOINT",
+		}, "")
+		if customEndpoint != nil {
+			data.FirebaserulesCustomEndpoint = types.StringValue(customEndpoint.(string))
+		}
+	}
+
+	if data.NetworkConnectivityCustomEndpoint.IsNull() {
+		customEndpoint := MultiEnvDefault([]string{
+			"GOOGLE_NETWORK_CONNECTIVITY_CUSTOM_ENDPOINT",
+		}, "")
+		if customEndpoint != nil {
+			data.NetworkConnectivityCustomEndpoint = types.StringValue(customEndpoint.(string))
+		}
+	}
+
+	if data.RecaptchaEnterpriseCustomEndpoint.IsNull() {
+		customEndpoint := MultiEnvDefault([]string{
+			"GOOGLE_RECAPTCHA_ENTERPRISE_CUSTOM_ENDPOINT",
+		}, "")
+		if customEndpoint != nil {
+			data.RecaptchaEnterpriseCustomEndpoint = types.StringValue(customEndpoint.(string))
+		}
+	}
+}
+
+func (p *frameworkProvider) SetupClient(ctx context.Context, data ProviderModel, diags *diag.Diagnostics) {
+	tokenSource := GetTokenSource(ctx, data, false, diags)
+	if diags.HasError() {
+		return
+	}
+
+	cleanCtx := context.WithValue(ctx, oauth2.HTTPClient, cleanhttp.DefaultClient())
+
+	// 1. MTLS TRANSPORT/CLIENT - sets up proper auth headers
+	client, _, err := transport.NewHTTPClient(cleanCtx, option.WithTokenSource(tokenSource))
+	if err != nil {
+		diags.AddError("error creating new http client", err.Error())
+		return
+	}
+
+	// Userinfo is fetched before request logging is enabled to reduce additional noise.
+	p.logGoogleIdentities(ctx, data, diags)
+	if diags.HasError() {
+		return
+	}
+
+	// 2. Logging Transport - ensure we log HTTP requests to GCP APIs.
+	loggingTransport := logging.NewTransport("Google", client.Transport)
+
+	// 3. Retry Transport - retries common temporary errors
+	// Keep order for wrapping logging so we log each retried request as well.
+	// This value should be used if needed to create shallow copies with additional retry predicates.
+	// See ClientWithAdditionalRetries
+	retryTransport := NewTransportWithDefaultRetries(loggingTransport)
+
+	// 4. Header Transport - outer wrapper to inject additional headers we want to apply
+	// before making requests
+	headerTransport := newTransportWithHeaders(retryTransport)
+	if !data.RequestReason.IsNull() {
+		headerTransport.Set("X-Goog-Request-Reason", data.RequestReason.ValueString())
+	}
+
+	// Ensure $userProject is set for all HTTP requests using the client if specified by the provider config
+	// See https://cloud.google.com/apis/docs/system-parameters
+	if data.UserProjectOverride.ValueBool() && !data.BillingProject.IsNull() {
```

Comment:
I think this might make a slightly different check than we think? IIUC `!data.BillingProject.IsNull()` doesn't mean the same thing as `c.BillingProject != ""`.
We set `data.BillingProject with the following: `data.BillingProject = types.StringValue(os.Getenv("GOOGLE_BILLING_PROJECT"))`.

`StringValue()` calls [NewStringValue](https://github.com/hashicorp/terraform-plugin-framework/blob/f906a5983b9ce9c4d61720aa7ac2db1c8a2b6ed7/types/basetypes/string_value.go#L53) which sets the string state to attr.ValueStateKnown. [IsNull](https://github.com/hashicorp/terraform-plugin-framework/blob/f906a5983b9ce9c4d61720aa7ac2db1c8a2b6ed7/types/basetypes/string_value.go#L123) doesn't check the value ("") but instead checks the state to see if it's attr.ValueStateNull.

That means we'll pass this condition when `user_project_override = true` and no `billing_project` is set, or `billing_project = ""`, setting the header with a blank target project. Previously, both those cases would fail.

Reply:
Oh nevermind. This is handled by the following (L182 at time of writing), we moved the check up.

```
if data.RequestReason.IsNull() && os.Getenv("CLOUDSDK_CORE_REQUEST_REASON") != "" {
		data.RequestReason = types.StringValue(os.Getenv("CLOUDSDK_CORE_REQUEST_REASON"))
	}
```

Feel free to resolve, leaving this comment up so I can refer back the next time a null trips me up